tests/components/ajax/aggregate.test.js

```diff
@@ -0,0 +1,245 @@
+        undefined, FEATURE_NAMES.softNav, expect.any(Object))
+
+      softNavInUse = false
+    })
+  })
+
+  test('on interactionDiscarded, saved (old) SPA events are put back in ajaxEvents', () => {
+    const interaction = { id: 0 }
+    context.spaNode = { interaction }
+
+    ajaxAggregate.ee.emit('xhr', ajaxArguments, context)
+    ajaxAggregate.ee.emit('interactionDone', [interaction, false])
+
+    expect(ajaxAggregate.spaAjaxEvents[interaction.id]).toBeUndefined() // no interactions in SPA under interaction 0
+    expect(ajaxAggregate.ajaxEvents.length).toEqual(1)
+
+    ajaxAggregate.ajaxEvents = []
+  })
+  test('on returnAjax from soft nav, event is re-routed back into ajaxEvents', () => {
+    softNavInUse = true
+
+    ajaxAggregate.ee.emit('xhr', ajaxArguments, context)
+    const event = hMod.handle.mock.lastCall[1][0]
+    ajaxAggregate.ee.emit('returnAjax', [event], context)
+
+    expect(ajaxAggregate.ajaxEvents.length).toEqual(1)
+    expect(ajaxAggregate.ajaxEvents[0]).toEqual(expect.objectContaining({ startTime: 0, path: '/pathname' }))
+
+    softNavInUse = false
+  })
+
+  describe('prepareHarvest', () => {
+    test('correctly serializes an AjaxRequest events payload', () => {
+      const expected = {
+        type: 'ajax',
+        start: 0,
+        end: 30,
+        callbackEnd: 30,
+        callbackDuration: 0,
+        domain: 'https://example.com',
+        path: '/pathname',
+        method: 'PUT',
+        status: 200,
+        requestedWith: 'XMLHttpRequest',
+        requestBodySize: 128,
+        responseBodySize: 256,
+        nodeId: '0',
+        guid: null,
+        traceId: null,
+        timestamp: null
+      }
+
+      ajaxAggregate.ee.emit('xhr', ajaxArguments, context)
+
+      const expectedCustomAttributes = {
+        customStringAttribute: 'customStringAttribute',
+        customNumAttribute: 2,
+        customBooleanAttribute: true,
+        nullCustomAttribute: null,
+        undefinedCustomAttribute: undefined // will be treated as null
+      }
+      mockCurrentInfo.jsAttributes = expectedCustomAttributes
+
+      const serializedPayload = ajaxAggregate.prepareHarvest({ retry: false })
+      // serializedPayload from ajax comes back as an array of bodies now, so we just need to decode each one and flatten
+      // this decoding does not happen elsewhere in the app so this only needs to happen here in this specific test
+      const decodedEvents = serializedPayload.map(sp => qp.decode(sp.body.e))
+
+      decodedEvents.forEach(payload => {
+        payload.forEach(event => {
+          validateCustomAttributeValues(expectedCustomAttributes, event.children)
+          delete event.children
+
+          expect(event).toEqual(expected) // event attributes serialized correctly
+        })
+      })
+
+      mockCurrentInfo.jsAttributes = {} // reset jsattributes for other tests
```

Comment:
can we put stuff like this in the afterEach block so its more clear what sideEffects we are managing through the tests